linalg/sparsemat.cpp

```diff
@@ -580,23 +609,78 @@ void SparseMatrix::AddMult(const Vector &x, Vector &y, const double a) const
    }
 
 #ifndef MFEM_USE_LEGACY_OPENMP
+
    const int height = this->height;
    const int nnz = J.Capacity();
    auto d_I = Read(I, height+1);
    auto d_J = Read(J, nnz);
    auto d_A = Read(A, nnz);
    auto d_x = x.Read();
    auto d_y = y.ReadWrite();
-   MFEM_FORALL(i, height,
+
+   //Skip if matrix is all zero
+   if (nnz == 0) {return;}
+#ifdef MFEM_USE_CUDA
+   if (Device::Allows(Backend::CUDA_MASK))
    {
-      double d = 0.0;
-      const int end = d_I[i+1];
-      for (int j = d_I[i]; j < end; j++)
+
+
+
+      const double alpha = a;
+      const double beta  = 1.0;
+
+      //Initialize once
+      if (!isInit)
       {
-         d += d_A[j] * d_x[d_J[j]];
+         /* create and setup matrix descriptor */
+         cusparseCreateCsr(&matA_descr,Height(), Width(), J.Capacity(),
+                           const_cast<int *>(d_I),
+                           const_cast<int *>(d_J), const_cast<double *>(d_A), CUSPARSE_INDEX_32I,
+                           CUSPARSE_INDEX_32I, CUSPARSE_INDEX_BASE_ZERO, CUDA_R_64F);
+
```

Comment:
Does cuSPARSE just wrap the data and not use any additional GPU memory?

Reply:
I believe it does use a bit more of GPU memory, there is a step below where it allocates some extra memory.